panic: runtime error: index out of range in the ingester with native histograms

#### Describe the bug
Ingesting native histograms can cause a panic in `counterResetInAnyBucket` which then corrupts the WAL.

#### To Reproduce
Steps to reproduce the behavior:

1. Start Mimir ([r248-c7fbeb7](https://hub.docker.com/layers/grafana/mimir/r248-c7fbeb7/images/sha256-2b0466138fad258a445bb8a7cd271e5df3088afefdc87f5d0bfaa7b14f50fb75?context=explore))
2. Start OTEL Collector ([8a4348cb009d1448dc647f2bdad27304b8ce53d3](https://hub.docker.com/layers/otel/opentelemetry-collector-contrib-dev/8a4348cb009d1448dc647f2bdad27304b8ce53d3/images/sha256-f086dafb0545da52d8ede77d5a355dafb63c2bb1ad9ff801245abca672d6fe66?context=explore))
3. Generate exponential histograms via OTEL client libs, push these to the collector and have the collector write them to Mimir via the `prometheusremotewrite` exporter as native histograms.

#### Expected behavior

No panics.

#### Environment

- Infrastructure: kubernetes
- Deployment tool: helm

#### Additional Context

I decided to kick the tires on the histogram downsampling (from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/24026), as we'd like to start using these from the OTEL libs. Reconfiguring the collector to write downsampled native histograms immediately led to the below error in the mimir ingesters. After restarting I then got the same error during WAL replay. Ended up having to wipe the WALs completely. I see there has been a previous PR for `counterResetInAnyBucket` which looks very similar in https://github.com/prometheus/prometheus/pull/12357.

Logs:

```
2023-07-24 10:37:12	panic: runtime error: index out of range [2] with length 2
2023-07-24 10:37:12
2023-07-24 10:37:12	goroutine 1212 [running]:
2023-07-24 10:37:12	github.com/prometheus/prometheus/tsdb/chunkenc.counterResetInAnyBucket({0x405595bd20?, 0xa624e0?, 0x4000e9bc48?}, {0x40584a9330?, 0x40584acd00?, 0x1898705c49c?}, {0x405595bd10?, 0x26e6ca0?, 0x400138e000?}, {0x40584a4558?, ...})
2023-07-24 10:37:12		/__w/mimir/mimir/vendor/github.com/prometheus/prometheus/tsdb/chunkenc/histogram.go:423 +0x130
2023-07-24 10:37:12	github.com/prometheus/prometheus/tsdb/chunkenc.(*HistogramAppender).Appendable(0x4056129c20, 0x40584d65a0)
2023-07-24 10:37:12		/__w/mimir/mimir/vendor/github.com/prometheus/prometheus/tsdb/chunkenc/histogram.go:309 +0x1c4
2023-07-24 10:37:12	github.com/prometheus/prometheus/tsdb.(*memSeries).appendHistogram(0x4055e5aea0, 0x1898705c49c, 0x40584d65a0, 0x0, {{0x26e6ca0?, 0x400138e000?}, 0x0?, 0x0?})
2023-07-24 10:37:12		/__w/mimir/mimir/vendor/github.com/prometheus/prometheus/tsdb/head_append.go:1197 +0x1d4
2023-07-24 10:37:12	github.com/prometheus/prometheus/tsdb.(*walSubsetProcessor).processWALSamples(0x40051ba158, 0x4000d6f000, 0x0?, 0x0?)
2023-07-24 10:37:12		/__w/mimir/mimir/vendor/github.com/prometheus/prometheus/tsdb/head_wal.go:627 +0x4e0
2023-07-24 10:37:12	github.com/prometheus/prometheus/tsdb.(*Head).loadWAL.func2(0x403ebf2d88?)
2023-07-24 10:37:12		/__w/mimir/mimir/vendor/github.com/prometheus/prometheus/tsdb/head_wal.go:106 +0x48
2023-07-24 10:37:12	created by github.com/prometheus/prometheus/tsdb.(*Head).loadWAL
2023-07-24 10:37:12		/__w/mimir/mimir/vendor/github.com/prometheus/prometheus/tsdb/head_wal.go:105 +0x23c
```

I've just built Prometheus from main and tried to load your WAL:
```ts=2023-09-13T09:55:38.993Z caller=main.go:206 level=info msg="Experimental native histogram support enabled."
ts=2023-09-13T09:55:38.993Z caller=main.go:539 level=info msg="No time or size retention was set so using the default time retention" duration=15d
ts=2023-09-13T09:55:38.993Z caller=main.go:583 level=info msg="Starting Prometheus Server" mode=server version="(version=2.47.0, branch=main, revision=4419399e4e831052bc9b8b4f26a8e7cf337091b0)"
ts=2023-09-13T09:55:38.993Z caller=main.go:588 level=info build_context="(go=go1.20.8, platform=linux/amd64, user=krajo@hpkrajo, date=20230913-09:47:33, tags=netgo,builtinassets,stringlabels)"
ts=2023-09-13T09:55:38.993Z caller=main.go:589 level=info host_details="(Linux 5.15.0-82-generic #91~20.04.1-Ubuntu SMP Fri Aug 18 16:24:39 UTC 2023 x86_64 hpkrajo (none))"
ts=2023-09-13T09:55:38.993Z caller=main.go:590 level=info fd_limits="(soft=1048576, hard=1048576)"
ts=2023-09-13T09:55:38.993Z caller=main.go:591 level=info vm_limits="(soft=unlimited, hard=unlimited)"
ts=2023-09-13T09:55:38.995Z caller=web.go:566 level=info component=web msg="Start listening for connections" address=0.0.0.0:9090
ts=2023-09-13T09:55:38.997Z caller=main.go:1024 level=info msg="Starting TSDB ..."
ts=2023-09-13T09:55:38.999Z caller=tls_config.go:274 level=info component=web msg="Listening on" address=[::]:9090
ts=2023-09-13T09:55:38.999Z caller=tls_config.go:277 level=info component=web msg="TLS is disabled." http2=false address=[::]:9090
ts=2023-09-13T09:55:39.003Z caller=head.go:600 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
ts=2023-09-13T09:55:39.003Z caller=head.go:681 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=4.26µs
ts=2023-09-13T09:55:39.003Z caller=head.go:689 level=info component=tsdb msg="Replaying WAL, this may take a while"
ts=2023-09-13T09:55:39.568Z caller=head.go:725 level=info component=tsdb msg="WAL checkpoint loaded"
ts=2023-09-13T09:55:40.507Z caller=head.go:760 level=info component=tsdb msg="WAL segment loaded" segment=858 maxSegment=870
ts=2023-09-13T09:55:41.780Z caller=head.go:760 level=info component=tsdb msg="WAL segment loaded" segment=859 maxSegment=870
ts=2023-09-13T09:55:42.775Z caller=head.go:760 level=info component=tsdb msg="WAL segment loaded" segment=860 maxSegment=870
panic: runtime error: index out of range [26] with length 26

goroutine 494 [running]:
github.com/prometheus/prometheus/tsdb/chunkenc.counterResetInAnyBucket({0xc01625d380?, 0x0?, 0x7f05fd752d00?}, {0xc014367860, 0x1a, 0x80?}, {0xc0172a0c18, 0x3, 0x0?}, {0xc01590c2b0, ...})
	/home/krajo/go/github.com/krajorama/prometheus/tsdb/chunkenc/histogram.go:394 +0x23e
github.com/prometheus/prometheus/tsdb/chunkenc.(*HistogramAppender).appendable(0xc00a7f8000, 0xc017699b00)
	/home/krajo/go/github.com/krajorama/prometheus/tsdb/chunkenc/histogram.go:300 +0x225
github.com/prometheus/prometheus/tsdb/chunkenc.(*HistogramAppender).AppendHistogram(0xc00a7f8000, 0x18a8ad99576?, 0xc008f54f02?, 0xc017699b00, 0x0)
	/home/krajo/go/github.com/krajorama/prometheus/tsdb/chunkenc/histogram.go:637 +0x8e
github.com/prometheus/prometheus/tsdb.(*memSeries).appendHistogram(0xc007ed5380, 0x18a8ad99576, 0xc017699b00, 0x0, {0xc0001985a0?, 0x0?, 0x0?})
	/home/krajo/go/github.com/krajorama/prometheus/tsdb/head_append.go:1178 +0xdd
github.com/prometheus/prometheus/tsdb.(*walSubsetProcessor).processWALSamples(0xc00e631848, 0xc000053800, 0x0?, 0x0?)
	/home/krajo/go/github.com/krajorama/prometheus/tsdb/head_wal.go:623 +0x675
github.com/prometheus/prometheus/tsdb.(*Head).loadWAL.func2(0x0?)
	/home/krajo/go/github.com/krajorama/prometheus/tsdb/head_wal.go:106 +0x4d
created by github.com/prometheus/prometheus/tsdb.(*Head).loadWAL
	/home/krajo/go/github.com/krajorama/prometheus/tsdb/head_wal.go:105 +0x31f
```

As suspected the bug is in Prometheus, not Mimir, I'm opening an issue in Prometheus to link this to.

--- a/go.mod
+++ b/go.mod
@@ -248,7 +248,7 @@ require (
 )
 
 // Using a fork of Prometheus with Mimir-specific changes.
-replace github.com/prometheus/prometheus => github.com/grafana/mimir-prometheus v0.0.0-20230922093515-d3ab1ce5060a
+replace github.com/prometheus/prometheus => github.com/grafana/mimir-prometheus v0.0.0-20230922140437-a10cef685b49
 
 // Replace memberlist with our fork which includes some fixes that haven't been
 // merged upstream yet:

